<a href="https://colab.research.google.com/github/Roseco-crs/bstat/blob/main/Neural_Network_from_scratch.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Implementation of Neural Network from scratch

In [42]:
# import Pytorch library
import torch

In [37]:
# Input 
X = torch.Tensor([[1,0,1,0], [1,0,1,1], [0,1,0,1]])
# Output
y = torch.Tensor([[1], [1], [0]])
print(f"Input: \n{X} \nOutput: \n{y}")

Input: 
tensor([[1., 0., 1., 0.],
        [1., 0., 1., 1.],
        [0., 1., 0., 1.]]) 
Output: 
tensor([[1.],
        [1.],
        [0.]])


Definition of activation function.
We will use sigmoid function for forward propagation and derivative of sigmoid for backward propagation

In [38]:
def sigmoid(t):
  return 1/(1+torch.exp(-t))

def derivative_sigmoid(t):
  sig = sigmoid(t)
  return sig*(1-sig)
  

Neural Network model architecture:

Initialization of 

1) parameters of the model 

2) hyperparameters such as epoch and learning rate

In [39]:
""" Neural Network model architecture : 
    1) input
    2) a single hidden layer
    3) output
"""

inputlayer_neurons = X.shape[1]    # number of features in data  X. In our case it is 4
hiddenlayer_neurons = 5            # number of hidden layer. We choose 5
output_neurons = y.shape[1]        # number of output. In our case it is 1

print(f"X shape: {X.shape} \ny shape: {y.shape} \nFeatures in data X: {X.shape[1]} \nFeatures in output: {y.shape[1]}")

# Weights and Bias initialization
input_hidden_weight = torch.randn(inputlayer_neurons, hiddenlayer_neurons).type(torch.FloatTensor)
input_hidden_bias = torch.randn(1, hiddenlayer_neurons).type(torch.FloatTensor)
hidden_output_weight = torch.randn(hiddenlayer_neurons, output_neurons)
hidden_output_bias = torch.randn(1, output_neurons)

# hyperparameters
epoch = 5000                        # Training iterations
lr = 0.3                            # Learning rate


X shape: torch.Size([3, 4]) 
y shape: torch.Size([3, 1]) 
Features in data X: 4 
Features in output: 1


Build the model


In [40]:
for i in range(epoch):
  #Forward Propagation
  hidden_layer_input1 = torch.mm(X, input_hidden_weight)
  hidden_layer_input = hidden_layer_input1 + input_hidden_bias
  hidden_layer_activation = sigmoid(hidden_layer_input)
  
  output_layer_input1 = torch.mm(hidden_layer_activation, hidden_output_weight)
  output_layer_input = output_layer_input1 + hidden_output_bias
  output = sigmoid(output_layer_input)
  
  #Backward Propagation
  error = y-output
  slope_output_layer = derivative_sigmoid(output)
  slope_hidden_layer = derivative_sigmoid(hidden_layer_activation)
  d_output = error * slope_output_layer
  error_at_hidden_layer = torch.mm(d_output, hidden_output_weight.t())
  d_hiddenlayer = error_at_hidden_layer * slope_hidden_layer


  hidden_output_weight += torch.mm(hidden_layer_activation.t(), d_output)*lr
  hidden_output_bias += d_output.sum()*lr
  input_hidden_weight += torch.mm(X.t(), d_hiddenlayer)*lr
  input_hidden_bias += d_output.sum()*lr
  

In the forward propagation we calculate the output.

In the backward propagation we calculate the error and update the weights and biases using this error.

Let's see the output from the model.

In [41]:
print(f"Acutal :\n{y},\nPredicted :\n{output}")

Acutal :
tensor([[1.],
        [1.],
        [0.]]),
Predicted :
tensor([[0.9993],
        [0.9991],
        [0.0013]])


The target is 1,1,0 and the predicted values from our model are 0.9993, 0.9991, 0.0013. Not bad at all!

This is how we can build and train a neural network from strach in Pytorch.